### ISSUE TITLE
cyclic terms are rendered incompletely

### ISSUE THREAD

**Issue by jburse** [jariazavalverde/tau-prolog#75]
Tau Prolog shows a behaviour which is somewhere between 
GNU Prolog (does not hang) and SWI-Prolog (makes full factoring).

I get the following result:

**Comment by jburse** [comment]
On the other in SWI-Prolog we get full factoring of the cyclic term:

![image](https://user-images.githubusercontent.com/5411487/53305608-d06a2800-3883-11e9-9ead-ee1dbafd40a6.png)

**Comment by jariazavalverde** [comment]
In Tau Prolog, in each resolution step a substitution is created and composed with the substitution of the previous state, without taking into account that a cyclic term can be generated.

For example, for the following goal:
`X = f(Y), Y = g(Y).`

Tau Prolog starts with an initial substitution `{X/X,Y/Y}` and then:
`<[X=f(Y), Y=g(Y)], {X/X,Y/Y}> =>`
`<[Y=g(Y)], {X/f(Y),Y/Y}> =>`
`<[], {X/f(g(Y)),Y/g(Y)}>`

**Comment by jburse** [comment]
I think the toplevel display can not distingish the following two programs in their answer substitution:

test1(X) :- X = f(Y),  Y = f(_).

test2(X) :- X = f(Y),  Y = f(Y).

I get the following result:

**Comment by jburse** [comment]
Now compare with SWI-Prolog:

![image](https://user-images.githubusercontent.com/5411487/53306221-a10ae980-388a-11e9-9621-baf6994ba77f.png)

**Comment by jariazavalverde** [comment]
But note that Tau Prolog can detect if there is an acyclic term or not, so it's not really a problem.

![goal](https://user-images.githubusercontent.com/5055295/53306237-e3342b00-388a-11e9-89e3-4a5924cd528c.png)

**Comment by jburse** [comment]
Yes. I only wondered what other Prolog systems do in the top-level. acyclic_term/1 is specified in the ISO core standard. The top-level not specified in the ISO core standard. So its more a comfort feature request, not a bug versus some standard.

**Comment by jariazavalverde** [comment]
I think it's computationally expensive for Tau Prolog to detect by default cyclic terms in the answers (because it is not very frequent), but I guess I could create another method to print the cyclic terms as you want.

**Comment by jburse** [comment]
I think you detect already cyclic terms. I find:
